### PR TITLE
compaction_manager: rewrite_sstables: do not acquire table write lock

### LINKED ISSUE #10175: Cleanup compaction / rewrite sstable block behind regular compaction for each sstable
Opening this issue based on my analysis in https://github.com/scylladb/scylla/issues/10060#issuecomment-1059951822

With 40a6049ac285f1a317c6c804c09ae7363cd73ffb, rewrite_sstables takes the write lock for each sstable.
Since it will block behind regular compaction, the whole cleanup task might take a very long time.

The purpose of this change was to serialize cleanup or upgrade_sstables with other "maintenance" compaction tasks.

This was achieved in a better way in 6737c880458c216c9ab933b0348b9f78fe79c648, and after this change rewrite_sstables can safely take the read lock rather the write lock, and let regular compaction run in parallel without interruption.

That said, I don't understand why rewrite_sstables acquires the `_maintenance_ops_sem` per sstable and not once for the whole task.

### PR DESCRIPTION
**Pull request by bhalevy**
This change effectively reverts 40a6049ac285f1a317c6c804c09ae7363cd73ffb
that is no longer required after 6737c880458c216c9ab933b0348b9f78fe79c648
where the _maintenance_ops_sem is used to serialize
tasks using rewrite_sstables like cleanup, upgrade, reshape,
or scrub, with major compaction.
    
Since regular compaction may run in parallel no lock
is required per-table.
    
This solves the slowness in cleanup compaction (and
likely in upgrade sstables seen in #10060, and possibly #10166.

With that, rewrite_sstable can run compaction in the maintenance
scheduling group again, reinstating 4c05e5f966d691594a604cf076113efc7f42b1af and reverting a9427f150abd20227f8f967ae7b768480097b608.
    
Fixes #10175
Refs #10060

Test: unit(dev)
DTest: compaction_test.py compaction_additional_test.py

All passed except for currently failing dtests: test_double_compaction_by_cleanup_and_major_compactions (scylladb/scylla-dtest#2394) and test_twcs_multiple_sstables_during_decommission (scylladb/scylla-dtest#2616)

### CODE REVIEW

**Comment by bhalevy** [comment]
@roydahan can you please SCT with this branch to verify that time to cleanup indeed got back to pre- 4.6 times based on #10060?

Also, this fix should also address #10166, so I'd appreciate QA testing upgradesstables as well (although there the logs seem to indicate `nodetool upgradesstables` was executed with the `-a` or `--all` option.

File: compaction/compaction_manager.cc
Comment (raphaelsc) [inline compaction/compaction_manager.cc:816]: what about taking read lock instead? it's not needed now, but to make it more future proof, e.g. guarantee serialization with a task type that takes write lock but doesn't go through maintenance_ops semaphore.

File: compaction/compaction_manager.cc
Comment (raphaelsc) [inline compaction/compaction_manager.cc:805]: I don't know if running rewrite_sstables() in maintenance group is a good idea.

That's because sstables locked by it influence compaction shares.

If compaction shares are increasing significantly, as new sstables arrive, we'd like to be able to speed up cleanup to make sure backlog can be decreased. Making this change is decoupling the controller from any maintenance operation.

Comment (bhalevy) [reply compaction/compaction_manager.cc:805]: We can separate this change to let this PR be merged earlier as a first step and examine running rewrite_sstables in maintenance sg as a follow-up, maybe with additional improvements to address any issues caused by it.

For example, we could implement a controller for maintenance scheduling group shares that could take the backlog into account as well as other inputs.  In essence we should decouple maintenance scheduling from regular compaction scheduling.  They are serving different purposes.

As for cleanup to complete faster due to backlog, new sstables should not need cleanup as they are written after the respective topology change(s) are already done, so they shouldn't contain any tokens not belonging to the node. The reason to speed up cleanup to reduce the backlog is that it prevents regular compaction from selecting sstables before they are cleaned up.

If we had a `requires_cleanup` flag on the sstable we could mark them when selected for cleanup at the beginning of the cleanup task but not register them as compacting, so that regular compaction can select them too and apply the cleanup filter when reading the sstable as part of regular compaction. When cleanup gets to processing the sstables, it can register them one at a time, as they are processed, to prevent regular compaction from accessing them concurrently.  If the sstable was already compacted and is now unlinked, cleanup can just skip it.

File: compaction/compaction_manager.cc
Comment (bhalevy) [reply compaction/compaction_manager.cc:816]: I strongly oppose it.  Although acquiring the read lock would logically work, cpu cycles aren't free.
It's dead weight that makes the code slightly slower and makes it harder to understand, and it can be easily re-introduced if and when needed.

The code would be much easier to understand where each statement and expression has a clear purpose, vs. having to rely on uncertain future-proof plans to figure out why something like a lock is taken with no real need.

Comment (bhalevy) [reply compaction/compaction_manager.cc:816]: You know what, I'll change it to a read lock for backporting purposes, but will get rid of it and all other `_compaction_state[t]` locks below maintenance_ops_sem other than major since it's really used only to synchronize major and regular compaction per-table.

**Comment by bhalevy** [comment]
In v2 (a97af5715a89816d97d481989c56d4370ab2317e):
- left only the removal of the per-table write lock in rewrite_sstables
  - turned into a read-lock to reduce backporting risk (if case the target branch, e.g. 4.5 doesn't have 6737c880458c216c9ab933b0348b9f78fe79c648.
- using maintenance scheduling group deferred to a later time.

File: compaction/compaction_manager.cc
Comment (raphaelsc) [reply compaction/compaction_manager.cc:816]: Fair enough. That's indeed the reason the rwlock was introduced for.

**Review by raphaelsc — APPROVED** [review]
Looks good to me.

**Comment by bhalevy** [comment]
@scylladb/scylla-maint please merge.

I understand from @raphaelsc that after re-analyzing the logs in #10060 he concluded that out of 19h of cleanup on all nodes, about 16 can be attributed to delays in cleanup compaction where it is blocked behind regular compaction, waiting to acquire the per-table write-lock, that this PR fixes.

Cc @slivne

**Comment by roydahan** [comment]
@bhalevy so, do we need to retest with the above packages or not?

**Comment by bhalevy** [comment]
> @bhalevy so, do we need to retest with the above packages or not?

We can wait until it's merged and test packages from master

**Review by denesb — APPROVED** [review]

**Comment by slivne** [comment]
@fgelcer - need to look at the jenkins run and update the results of how long did cleanup take

**Comment by fgelcer** [comment]
> @fgelcer - need to look at the jenkins run and update the results of how long did cleanup take

https://github.com/scylladb/scylla/issues/10060#issuecomment-1087436893

**Pull request closed without merging**
